starknet_devnet/predeployed_contract_wrapper.py

```diff
@@ -1,9 +1,100 @@
 """Parent for predeployed contract wrapper classes"""
+from abc import ABC, abstractmethod
 
+from starkware.starknet.testing.starknet import Starknet
+from starkware.starknet.services.api.contract_class import ContractClass
+from starkware.starknet.testing.contract import StarknetContract
 
-class PredeployedContractWrapper:
+
+class PredeployedContractWrapper(ABC):
     """Parent class for all predeployed contract wrapper classes"""
 
+    # Cannot import it because of circular imports
+    # from .starknet_wrapper import StarknetWrapper
+
+    starknet_wrapper = NotImplemented
+
+    @property
+    def address(self) -> int:
+        try:
+            return self.ADDRESS
+        except AttributeError:
+            try:
+                return self._address
+            except AttributeError:
+                raise NotImplementedError(
+                    "Predeployed contracts should provide an address attribute or an ADDRESS class attribute"
+                )
+
+    @address.setter
+    def address(self, address: int):
+        self._address = address
+
+    @property
+    def class_hash_bytes(self) -> bytes:
+        try:
+            return self._class_hash_bytes
+        except AttributeError:
+            try:
+                return self.HASH_BYTES
+            except AttributeError:
+                raise NotImplementedError(
+                    "Predeployed contracts should provide a class_hash_bytes attribute or a HASH_BYTES class attribute"
+                )
+
+    @class_hash_bytes.setter
+    def class_hash_bytes(self, class_hash_bytes: bytes):
+        self._class_hash_bytes = class_hash_bytes
+
+    @property
+    def contract_class(self) -> ContractClass:
+        try:
+            return self._contract_class
+        except AttributeError:
+            try:
+                return self.get_contract_class()
+            except AttributeError:
+                raise NotImplementedError(
+                    "Predeployed contracts should provide a contract_class attribute or get_contract_class method"
+                )
+
+    @contract_class.setter
+    def contract_class(self, contract_class):
+        self._contract_class = contract_class
+
+    async def deploy_contract(self):
```

Comment:
I don't like it that we have both `deploy_contract` and `deploy`. If some of these methods are supposed to be protected (only accessed within the class and its children), maybe they can be prefixed with an underscore (`_`).

Reply:
We can even get rid of the function, its there so that we can easily switch between calling `state.deploy_contract` and manipulating the cache

BTW, what do we do about it ? Since you prefer calling `state.deploy_contract` but it doesn't work, obviously, for Fee Token ?

Reply:
Can all predeployed contracts directly manipulate the underlying object instead of calling `deploy_contract`?

Also, sure I prefer calling public methods rather than modifying private properties which shouldn't be modified and which make me suppress the linter xD But if they don't work, they don't work.

Reply:
Yes.

Reply:
Ok, then I'd consider all of the predeployed contracts deploying so (without `state.deploy_contract`), especially if it will be done in just one place (which it should since we are introducing this common parent class).